### ISSUE TITLE
Scala for Machine Learning 试译

### ISSUE THREAD

**Issue by satansk** [satansk/blog-gitbook-sk#16]
# 作者简介

Patrick R. Nicolas is the director of engineering at Agile SDE, California. He has more than 25 years of experience in software engineering and building applications in C++, Java, and more recently in Scala/Spark, and has held several managerial positions. His interests include real-time analytics, modeling, and the development of nonlinear models.

Patrick R. Nicolas 是加利福尼亚州 Agile SDE 公司的工程（技术）总监，拥有超过 25 年的软件工程经验，在过去 Patrick 使用 C++ 和 Java 构建应用，最近开始使用 Scala 和 Spark，除此之外，Patrick 拥有多个管理职位。Patrick 的兴趣包括实时分析、建模以及开发非线性模型。

# About the Reviewers

Sumit Pal has more than 24 years of experience in the software industry, spanning companies from start-ups to enterprises.

Sumit Pal 在软件工业界具有超过 24 年的经验，服务过从创业公司到商业公司。

He is a big data architect, visualization, and data science consultant, and builds endto-end data-driven analytic systems.

Sumit 是大数据架构师，可视化与数据科学咨询师，构建端到端的数据驱动的分析系统。

Sumit has worked for Microsoft (SQLServer), Oracle (OLAP), and Verizon (big data analytics).

Sumit 服务过微软（SQLServer）、Oracle（OLAP）和 Verizon（大数据分析）

Currently, he works for multiple clients, building their data architectures and big data solutions and works with Spark, Scala, Java, and Python.

目前，Sumit 服务于多个客户，帮助他们构建数据架构和大数据解决方案，Sumit 目前使用 Spark、Scala、Java 和 Python。

He has extensive experience in building scalable systems in middle tier, data tier to visualization for analytics applications, using big data and NoSQL databases.

从分析型应用的中间层、数据层到可视化，Sumit 具备丰富的使用大数据与 NoSQL 数据库为其构建可伸缩系统的经验。

Sumit has expertise in database internals, data warehouses, and dimensional modeling, as an associate director for big data at Verizon. Sumit strategized, managed, architected, and developed analytic platforms for machine learning applications. Sumit was the chief architect at ModelN/LeapfrogRX (2006-2013), where he architected the core analytics platform.

作为 Verizon 公司的大数据副主管，Sumit 在数据内部原理、数据仓库、维度建模等领域具备专家级经验。Sumit 曾负责机器学习应用分析平台的策略制定、管理、架构以及开发工作。Sumit 曾经是 ModelN/LeapfrogRX (2006-2013) 的首席架构师，负责核心分析平台的架构工作。

He is the author of SQL On Big Data - Technology, Architecture and Roadmap published by Apress in October 2016.

Sumit 是  SQL On Big Data - Technology, Architecture and Roadmap 一书的作者，该书由 Apress 与 2016 年出版。

He has spoken on the topic covered in this book at the following conferences:

• May 2016, Big Data Conference—Linux Foundation in Vancouver, Canada
• March 2016, World Data Center Conference in Las Vegas, USA
• November 2015, BigData TechCon in Chicago, USA
• August 2015, Global Big Data Conference in Boston, USA

Sumit 曾经就本书包含的主题在以下会议发表过演讲：

Dave Wentzel is the Chief Technology Officer (CTO) of Capax Global, a premier Microsoft consulting partner. Dave is responsible for setting the strategy and defining service offerings and capabilities for the data platform and Azure practice at Capax. Dave also works directly with clients to help them with their big data journey. Dave is a frequent blogger and speaker on big data and data science topics.

Dave Wentzel 是 Capax Global 公司的 CTO，首席微软咨询合伙人。

# 第二章 数据流水线

>In the first chapter, you were acquainted with some rudimentary concepts regarding data processing, clustering, and classification.

在第一章中，你已经熟悉了数据处理、聚类、分类的基本概念。

>This chapter is dedicated to the creation and maintenance of a flexible end-to-end workflow to train and classify data. The first section of the chapter introduces a data-centric (functional) approach to create number crunching applications, followed by a description of a configurable workflow computation model. The chapter concludes with an overview of different model validation techniques.

本章将介绍如何创建、维护一个灵活的、端到端的工作流，并用其对数据进行训练与分类。本章第一部分介绍创建数值计算应用的基于数据的（函数式）的方法，接下来是对可配置工作流计算模型的介绍。最后以多种模型验证技术结尾。

>You will learn how to do the following:
>
>• Apply the concept of monadic design to create dynamic workflows
>• Leverage some of Scala's advanced patterns, such as the cake pattern, to build portable computational workflows
>• Take into account the bias-variance trade-off in selecting a model
>• Overcome overfitting in modeling
>• Break down data into training, test and validation sets
>• Implement model validation in Scala using precision, recall, and F score

你将学会以下内容：

* 

## 模型

>Data is the lifeline of any scientist, and the selection of data providers is critical in developing or evaluating any statistical inference or machine learning algorithm.

数据是所有科学家的生命线，在开发任何统计推断或机器学习算法的过程中，数据提供者的选择至关重要。

### 什么是模型？

>We briefly introduced the concept of a model in the Model categorization section in Chapter 1, Getting Started.

在第一章的 *模型分类* 中，我们简单介绍了模型的概念。

>What constitutes a model? Wikipedia provides a reasonably good definition of a model as understood by scientists [2:1]:

模型由什么组成呢？维基百科提供了很好的科学家眼里的模型定义：

>A scientific model seeks to represent empirical objects, phenomena, and physical processes in a logical and objective way.
>Models that are rendered in software allow scientists to leverage computational power to simulate, visualize, manipulate and gain intuition about the entity, phenomenon or process being represented.

科学模型试图通过逻辑客观的方式，表示经验对象、现象和物理过程。

通过软件中的模型，科学家可以利用计算能力模拟、可视化、操纵和获取关于模型表示的实体、现象或过程的直觉。

>In statistics and probabilistic theory, a model describes data that one might observe from a system to express any form of uncertainty and noise. A model allows us to infer rules, make predictions, and learn from data.

在统计学与概率论中，模型描述可能从系统中观测到的、用来表示任意形式的不确定和噪音的数据。我们可以使用模型推断规则，做出预测，从数据中学习。

>A model is composed of features, also known as attributes or variables, and a set of relations between those features. For instance, the model represented by the function f(x,y) = x.sin(2y) has two features x and y and a relation f. Those two features are assumed to be independent. If the model is subject to a constraint, such as f(x, y) < 20, for example, then the conditional independence is no longer valid.

模型由特征（也被称为属性或变量）和一组特征间的关系组成。例如，函数 `f(x,y) = x.sin(2y)` 表示的模型有 `x` 和 `y` 两个特征，以及他们之间的关系 `f`，其中 `x` 和 `y` 特征是相互独立的。若该模型存在约束，例如 `f(x, y) < 20`，则条件独立不再有效。

An astute Scala programmer would associate a model to a monoid for which the set is the group of observations and the operator is the function implementing the model.

有经验的 Scala 程序员会把模型关联到 monoid，其中集合是可观测物组成的群，操作符是实现模型的函数。

模型的形式各不相同：

* Parametric: This consists of functions and equations (for example, y =
sin(2t+w))

* 参数式：由函数与方程组成（例如 y = sin(2t+w)）

* 微分模型: This consists of ordinary and partial differential equations
(for example, dy = 2x.dx)

* 微分式：由微分方程与偏微分方程组成（例如 dy = 2x.dx）

* Probabilistic: This consists of probability distributions (for example, p(x|c) =
exp (k.logx – x)/x!)

* 概率式：由概率分布组成（例如 p(x|c) = exp (k.logx – x)/x!）

* Graphical: This consists of graphs that abstract out the conditional
independence between variables (for example, p(x,y|c) = p(x|c).p(y|c))

* 图模型：由表示变量之间条件独立的图组成（例如 p(x,y|c) = p(x|c).p(y|c)）

* Directed graphs: This consists of a temporal, spatial relationships (for
example, scheduler)

* 有向图：由时间、空间关系组成（例如调度器）

* Numerical method: This consists of computational method such as finite difference, finite elements or Newton-Raphson

* 数值法：由诸如有限差分法、有限元法或牛顿-拉夫逊法等计算方法组成

* Chemistry: This consists of formulas and components (for example, H2O, Fe + C12 = FeC13)

* 化学：由公式与化学成分组成（例如 H2O, Fe + C12 = FeC13）

* Taxonomy: This consists of a semantic definition and a relationship of concepts (for example, APG/Eudicots/Rosids/Huaceae/Malvales)

* 分类学：由语义定义与概念之间的关系组成（例如 APG/Eudicots/Rosids/Huaceae/Malvales）

* Grammar and lexicon: This consists of a syntactic representation of documents (for example, Scala programming language)

* 语法与词法：由文档的语法表示组成（例如 Scala 编程语言）

* Inference logic: This consists of rules (for example, IF (stock vol> 1.5 * average) AND rsi> 80 THEN ...)

* 推理逻辑：由规则组成（例如 IF (stock vol> 1.5 * average) AND rsi> 80 THEN ...）

### Model versus design（模型、设计对比）

The confusion between model and design is quite common in computer science, the reason being that these terms have different meanings for different people depending on the subject. The following metaphors should help with your understanding of these two concepts:

在计算机科学中，模型与设计常常令人困惑，因为根据主题不同，对于不同的人，它们有不同的含义。下面的比喻将有助于你对这两个概念的理解：

* Modeling: This is describing something you know. A model assumes, which becomes an assertion if proven correct (for example, the US population p increases by 1.2% a year, dp/dt= 1.012).
* Designing: This is manipulating representation for things you don't know. Designing can be regarded as the exploration phase of modeling (for example, what are the features that contribute to the growth of US population? Birth rate? Immigration? Economic conditions? Social policies?).

* 模型：模型描述了你已知的事物，被证明正确之后，则模型的假设将变为断言（例如美国人口 p 每年增长 1.2%，模型为 dp/dt = 1.012）。
* 设计：设计试图表现你未知的事物，可以将设计视为建模的探索阶段（例如，哪些特征促进了美国人口的增长？出生率？移民？经济状况？社会政策？）。

### Selecting features 特征选择

The selection of a model's features is the process of discovering and documenting the minimum set of variables required to build the model. Scientists assume that data contains many redundant or irrelevant features. Redundant features do not provide information already given by the selected features, and irrelevant features provide no useful information.

模型的特征选择是指发现、记录构造模型所需变量的最小集合的过程。科学家们假定数据包含很多冗余、无关的特征，其中冗余特征提供的信息已选特征都能提供，而无关特征无法提供有用信息。

A features selection consists of two consecutive steps:
1. Search for new feature subsets.
2. Evaluate these feature subsets using a scoring mechanism.

特征选择包含两个步骤：

1. 寻找新的特征子集。
2. 通过评分机制评估这些特征子集。

The process of evaluating each possible subset of features to find the one that maximizes the objective function or minimizes the error rate is computationally intractable for large datasets. A model with n features requires 2n-1 evaluations!

评估特征子集的目的是，找到能最大化目标函数或者最小化错误率的子集，这大数据集而言，这是很困难的。一个具有 n 个特征的模型需要进行 2^n-1 次评估。

### Extracting features 特征提取

An observation is a set of indirect measurements of hidden, also known as latent, variables, which may be noisy or contain a high degree of correlation and redundancies. Using raw observations in a classification task would very likely produce inaccurate results. Using all features in each observation also incurs a high computation cost.

观察是对隐藏或隐式变量间接度量的集合，这些变量可能充满噪声，或者包含极高的相关性和冗余。在分类任务中直接使用原始观察值，很容易产生不准确的结果。在每次观察中使用的全部特征则计算量巨大。

The purpose of feature extraction is to reduce the number of variables or dimensions of the model by eliminating redundant or irrelevant features. The features are extracted by transforming the original set of observations into a smaller set at the risk of losing some vital information embedded in the original set.

特征提取的目的是，通过消除冗余、无关特征，以减少模型所需变量的数目或维度。特征提取是通过将观察值的原始集合转化为一个更小集合实现的，转化过程可能会丢失原始集合中的重要信息。

### Defining a methodology 方法论定义

Let's start by clarifying the role of the data scientist, software engineer, and domain expert.

首先澄清下数据科学家、软件工程师和领域专家的角色。

A domain or subject-matter expert is a person with authoritative or credited expertise in a particular area or topic. A chemist is an expert in the domain of chemistry and possibly related fields.

领域或主题专家是具备特定领域或主题的权威专业知识的人，例如化学家是化学或相关领域的专家。

A data scientist solves problems related to data in a variety of fields such as biological sciences, health care, marketing, or finances. Data and text mining, signal processing, statistical analysis, and modeling using machine learning algorithms are some of the activities performed by a data scientist.

数据科学家解决数据相关的问题，可能涉及生物科学、健康护理、市场或金融等不同领域，数据科学家的工作内容包括数据与文本挖掘、信号处理、统计分析、机器学习建模等。

A software developer performs all the tasks related to creating software applications, including analysis, design, coding, testing, and deployment.

软件开发者实施创建软件应用的所有相关任务，包括分析、涉及、编码、测试和部署。

A data scientist has many options in selecting and implementing a classification or clustering algorithm.

在挑选、实现分类或聚类算法时，数据科学家有多种选择。

Firstly, a mathematical or statistical model is to be selected to extract knowledge from the raw input data or the output of a data upstream transformation. The selection of the model is constrained by the following parameters:

首先，选择某个数学或统计模型，是为了从原始输入数据或数据转换上游的输出中提取知识的，模型的选择受以下条件约束：

• Business requirements, such as accuracy of results or computation time
• Availability of training data, algorithms, and libraries
• Access to a domain or subject-matter expert, if needed

* 业务需求，例如结果精度或计算时间
* 可用的训练数据、算法和库
* 如果需要，是否有领域或主题专家

Secondly, the engineer has to select a computational and deployment framework suitable for the amount of data to be processed. The computational context is to be defined by the following parameters:

其次，根据要处理的数据量，工程师需要选择合适的计算、部署框架。计算上下文由以下条件定义：

• Available resources, such as machines, CPU, memory, or I/O bandwidth
• Implementation strategy, such as iterative versus recursive computation or caching
• Requirement for responsiveness of the overall process, such as duration of computation or display of intermediate results

* 可用资源，例如机器、CPU、内存和 I/O 带宽
* 实现策略，例如迭代计算还是递归计算，或者缓存
* 对整个过程响应的需求，例如计算、展示实时结果所需时间

Thirdly, a domain expert has to tag or label the observations in order to generate an accurate classifier.

第三，领域专家需要对观察结果打标签，以生成精确的分类器。

Finally, the model has to be validated against a reliable test dataset.

最后，使用可靠的测试数据集验证模型。

The following diagram illustrates the selection process to create a workflow:

下图展示了创建一个工作流的过程：

The parameters of a data transformation may need to be reconfigured according to the output of the upstream data transformation. Scala's higher-order functions are particularly suitable for implementing configurable data transformations.

根据上游数据转换的输出，数据转换的参数可能需要重新配置，Scala 的高阶函数尤其适合实现可配置的数据转换。
